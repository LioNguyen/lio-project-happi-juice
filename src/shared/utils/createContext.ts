import type React from "react";
import { createContext as originCreateContext, useContext } from "react";

function createContext<T>(): [React.Provider<T>, () => T, React.Context<T>] {
  const context = originCreateContext<T>({} as T);

  return [context.Provider, () => useContext(context), context];
}

export { createContext };
