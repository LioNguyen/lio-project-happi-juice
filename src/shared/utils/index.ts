export * from './lazyLoad'
export * from './tailwind'
