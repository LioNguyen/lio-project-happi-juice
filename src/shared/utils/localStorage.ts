/**
 * Save data to localStorage
 * @param key - The key under which the data will be stored
 * @param value - The data to store, can be any serializable value
 */
function setLocalStorage<T>(key: string, value: T): void {
  try {
    const serializedValue =
      typeof value === 'string' ? value : JSON.stringify(value)

    localStorage.setItem(key, serializedValue)
  } catch (error) {
    console.error(`Error saving ${key} to localStorage:`, error)
  }
}

/**
 * Retrieve data from localStorage
 * @param key - The key of the stored data
 * @returns The parsed data or null if not found or error occurs
 */
function getLocalStorage<T>(key: string): T | null {
  try {
    const serializedValue = localStorage.getItem(key)

    if (!serializedValue || serializedValue === 'undefined') {
      return null
    }

    // Try to parse JSON
    try {
      return JSON.parse(serializedValue) as T
    } catch {
      // Return as string if JSON parsing fails
      return serializedValue as T
    }
  } catch (error) {
    console.error(`Error reading ${key} from localStorage:`, error)
    return null
  }
}

/**
 * Remove data from localStorage
 * @param key - The key to remove from localStorage
 */
function removeLocalStorage(key: string): void {
  try {
    localStorage.removeItem(key)
  } catch (error) {
    console.error(`Error removing ${key} from localStorage:`, error)
  }
}

/**
 * Clear all data from localStorage
 */
function clearLocalStorage(): void {
  try {
    localStorage.clear()
  } catch (error) {
    console.error('Error clearing localStorage:', error)
  }
}

export {
  clearLocalStorage,
  getLocalStorage,
  removeLocalStorage,
  setLocalStorage,
}
