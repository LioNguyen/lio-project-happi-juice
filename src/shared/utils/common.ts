import { format, isValid, Locale, parseISO } from 'date-fns'
import { enUS, vi } from 'date-fns/locale'
import _ from 'lodash'

import { IOrderItem } from '@/domains/order'
import { ISelectOption } from '@/shared/types'

interface INumberFormatOptions {
  decimals?: number
  thousandSeparator?: string
  decimalSeparator?: string
}

// ==================== Date Utilities ====================
/**
 * Parse various date formats into Date object
 * @category Date
 * @param {Date | string | undefined} value - Input date
 * @returns {Date | undefined} Parsed Date object or undefined if invalid
 * @example
 * parseDate("2024-02-26") // Date object
 * parseDate("invalid") // undefined
 */
function parseDate(value: Date | string | undefined): Date | undefined {
  if (!value) return undefined
  if (value instanceof Date) return isValid(value) ? value : undefined

  try {
    // Try ISO string parsing first
    const date = parseISO(value)
    if (isValid(date)) return date

    // Fallback to direct Date creation
    const normalDate = new Date(value)
    return isValid(normalDate) ? normalDate : undefined
  } catch {
    return undefined
  }
}

/**
 * Format a date string according to specified format
 * @category Date
 * @param {string} date - Date string to format
 * @param {string} [dateFormat] - Output format (defaults to "yyyy-MM-dd")
 * @returns {string} Formatted date string
 * @example
 * formatDateTime("2024-02-26") // "2024-02-26"
 * formatDateTime("2024-02-26", "dd/MM/yyyy") // "26/02/2024"
 */
function formatDateTime(date: string, dateFormat?: string) {
  return format(date, dateFormat || 'yyyy-MM-dd')
}

const localeMap: Record<string, Locale> = {
  vi: vi,
  en: enUS,
}

// ==================== Number & Price Utilities ====================
/**
 * Format a number with customizable separators and decimals
 * @category Number
 * @param {number} value - Number to format
 * @param {INumberFormatOptions} [options] - Formatting options
 * @returns {string} Formatted number string
 */
const formatNumber = (
  value: number,
  options?: INumberFormatOptions,
): string => {
  const {
    decimals = 0,
    thousandSeparator = ',',
    decimalSeparator = '.',
  } = options || {}

  const parts = value.toFixed(decimals).split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, thousandSeparator)

  return parts.join(decimalSeparator)
}

/**
 * Formats the price with thousand separators using locale-specific formatting
 * @category Price
 * @param {string} price - Price string to format
 * @returns {string} Formatted price string
 */
const formatPrice = (price: string) => {
  return Number(price).toLocaleString()
}

// ==================== Array Utilities ====================
/**
 * Transforms an array of select options into a sorted, joined string
 * @category Array
 * @param {ISelectOption[]} array - Array of select options
 * @param {string} [key] - Key to extract from each option
 * @param {string} [separator] - Separator for joining values
 * @returns {string} Sorted and joined string
 */
const transformToSortedString = (
  array: ISelectOption[],
  key: string = 'value',
  separator: string = ',',
): string => {
  return _.chain(array).map(key).sortBy().join(separator).value()
}

// ==================== Order Utilities ====================
/**
 * Calculates the total price for an array of order items
 * @category Order
 * @param {IOrderItem[]} items - Array of order items
 * @returns {number} Total price
 */
const calculateTotalPrice = (items: IOrderItem[]) => {
  return items.reduce((sum, item) => sum + item.price * item.quantity, 0)
}

export {
  // Date utils
  parseDate,
  formatDateTime,
  localeMap,

  // Number & Price utils
  formatNumber,
  formatPrice,

  // Array utils
  transformToSortedString,
  // Order utils
  calculateTotalPrice,
}
