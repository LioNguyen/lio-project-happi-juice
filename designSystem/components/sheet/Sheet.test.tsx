import { screen } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";

import { renderWithTheme } from "@/shared/utils";
import {
  Sheet,
  SheetClose,
  SheetContent,
  SheetDescription,
  SheetFooter,
  SheetHeader,
  SheetOverlay,
  SheetTitle,
  SheetTrigger,
} from "./Sheet";

// Mock lucide-react
vi.mock("lucide-react", () => ({
  X: () => null,
}));

// Mock dialog
vi.mock("@radix-ui/react-dialog", () => ({
  __esModule: true,
  default: vi.fn(),
  Root: vi.fn(({ children }) => <div data-testid="sheet-root">{children}</div>),
  Trigger: vi.fn(({ children }) => <button data-testid="sheet-trigger">{children}</button>),
  Close: vi.fn(({ children, className }) => (
    <button data-testid="sheet-close-button" className={className}>
      {children}
    </button>
  )),
  Portal: vi.fn(({ children }) => <div data-testid="sheet-portal">{children}</div>),
  Overlay: vi.fn(({ className, ...props }) => <div data-testid="sheet-overlay" className={className} {...props} />),
  Content: vi.fn(({ className, children, ...props }) => (
    <div data-testid="sheet-content" className={className} {...props}>
      {children}
    </div>
  )),
  Title: vi.fn(({ className, children, ...props }) => (
    <div data-testid="sheet-title" className={className} {...props}>
      {children}
    </div>
  )),
  Description: vi.fn(({ className, children, ...props }) => (
    <div data-testid="sheet-description" className={className} {...props}>
      {children}
    </div>
  )),
}));

describe("Sheet Component", () => {
  const renderSheet = (props = {}) => {
    return renderWithTheme(
      <Sheet>
        <SheetTrigger>Open Sheet</SheetTrigger>
        <SheetContent {...props}>
          <SheetHeader>
            <SheetTitle>Sheet Title</SheetTitle>
            <SheetDescription>Sheet Description</SheetDescription>
          </SheetHeader>
          <div>Content</div>
          <SheetFooter>
            <SheetClose>Close</SheetClose>
          </SheetFooter>
        </SheetContent>
      </Sheet>,
    );
  };

  // Basic Rendering Tests
  it("should render all sheet components", () => {
    renderSheet();

    expect(screen.getByTestId("sheet-root")).toBeInTheDocument();
    expect(screen.getByTestId("sheet-trigger")).toBeInTheDocument();
    expect(screen.getByTestId("sheet-content")).toBeInTheDocument();
    expect(screen.getByTestId("sheet-title")).toBeInTheDocument();
    expect(screen.getByTestId("sheet-description")).toBeInTheDocument();
  });

  // Snapshot Tests
  describe("Sheet Snapshots", () => {
    it("should match snapshot with default props", () => {
      const { container } = renderSheet();
      expect(container).toMatchSnapshot();
    });

    it("should match snapshot with custom side", () => {
      const { container } = renderSheet({ side: "left" });
      expect(container).toMatchSnapshot();
    });

    it("should match snapshot with custom content", () => {
      const { container } = renderWithTheme(
        <Sheet>
          <SheetTrigger>Custom Trigger</SheetTrigger>
          <SheetContent>
            <SheetHeader>
              <SheetTitle>Custom Title</SheetTitle>
              <SheetDescription>Custom Description</SheetDescription>
            </SheetHeader>
            <div>Custom Content</div>
          </SheetContent>
        </Sheet>,
      );
      expect(container).toMatchSnapshot();
    });
  });

  // SheetOverlay Tests
  describe("SheetOverlay", () => {
    it("should render with correct default classes", () => {
      renderWithTheme(<SheetOverlay />);

      const overlay = screen.getByTestId("sheet-overlay");
      expect(overlay).toHaveClass(
        "fixed",
        "inset-0",
        "z-50",
        "data-[state=open]:animate-in",
        "data-[state=closed]:animate-out",
        "data-[state=closed]:fade-out-0",
        "data-[state=open]:fade-in-0",
      );
    });
  });

  // SheetContent Tests
  describe("SheetContent", () => {
    it.each([["top"], ["bottom"], ["left"], ["right"]])("should render with correct classes for %s side", side => {
      renderSheet({ side });

      const content = screen.getByTestId("sheet-content");
      expect(content).toHaveClass(
        "fixed",
        "z-50",
        "gap-4",
        "bg-background",
        "p-6",
        "shadow-lg",
        "transition",
        "ease-in-out",
      );
    });

    it("should render close button with correct classes", () => {
      renderSheet();

      const closeButton = screen.getAllByTestId("sheet-close-button")[0];

      expect(closeButton).toHaveClass("absolute", "right-4", "top-4", "rounded-sm", "opacity-70");
    });
  });

  // SheetHeader Tests
  describe("SheetHeader", () => {
    it("should render with correct default classes", () => {
      renderWithTheme(
        <SheetHeader data-testid="sheet-header">
          <div>Header Content</div>
        </SheetHeader>,
      );

      const header = screen.getByTestId("sheet-header");
      expect(header).toHaveClass("flex", "flex-col", "space-y-2", "text-center", "sm:text-left");
    });
  });

  // SheetFooter Tests
  describe("SheetFooter", () => {
    it("should render with correct default classes", () => {
      renderWithTheme(
        <SheetFooter data-testid="sheet-footer">
          <button>Action</button>
        </SheetFooter>,
      );

      const footer = screen.getByTestId("sheet-footer");
      expect(footer).toHaveClass("flex", "flex-col-reverse", "sm:flex-row", "sm:justify-end", "sm:space-x-2");
    });
  });

  // SheetTitle and SheetDescription Tests
  describe("SheetTitle and SheetDescription", () => {
    it("should render title with correct classes", () => {
      renderWithTheme(<SheetTitle>Test Title</SheetTitle>);

      const title = screen.getByTestId("sheet-title");
      expect(title).toHaveClass("text-lg", "font-semibold", "text-foreground");
    });

    it("should render description with correct classes", () => {
      renderWithTheme(<SheetDescription>Test Description</SheetDescription>);

      const description = screen.getByTestId("sheet-description");
      expect(description).toHaveClass("text-sm", "text-muted-foreground");
    });
  });

  // Accessibility Tests
  describe("Accessibility", () => {
    it("should have close button with sr-only text", () => {
      renderSheet();

      const closeButton = screen.getAllByTestId("sheet-close-button")[0];
      expect(closeButton).toHaveTextContent("Close");
    });
  });
});
