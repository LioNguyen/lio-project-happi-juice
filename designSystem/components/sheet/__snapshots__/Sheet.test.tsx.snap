// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sheet Component > Sheet Snapshots > should match snapshot with custom content 1`] = `
<div>
  <div
    data-testid="sheet-root"
  >
    <button
      data-testid="sheet-trigger"
    >
      Custom Trigger
    </button>
    <div
      data-testid="sheet-portal"
    >
      <div
        class="fixed inset-0 z-50 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0"
        data-testid="sheet-overlay"
      />
      <div
        class="fixed z-50 gap-4 bg-background p-6 shadow-lg transition ease-in-out data-[state=closed]:duration-300 data-[state=open]:duration-500 data-[state=open]:animate-in data-[state=closed]:animate-out inset-y-0 right-0 h-full w-3/4 border-l data-[state=closed]:slide-out-to-right data-[state=open]:slide-in-from-right sm:max-w-sm"
        data-testid="sheet-content"
      >
        <button
          class="absolute right-4 top-4 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none  disabled:pointer-events-none data-[state=open]:bg-secondary"
          data-testid="sheet-close-button"
        >
          <span
            class="sr-only"
          >
            Close
          </span>
        </button>
        <div
          class="flex flex-col space-y-2 text-center sm:text-left"
        >
          <div
            class="text-lg font-semibold text-foreground"
            data-testid="sheet-title"
          >
            Custom Title
          </div>
          <div
            class="text-sm text-muted-foreground"
            data-testid="sheet-description"
          >
            Custom Description
          </div>
        </div>
        <div>
          Custom Content
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Sheet Component > Sheet Snapshots > should match snapshot with custom side 1`] = `
<div>
  <div
    data-testid="sheet-root"
  >
    <button
      data-testid="sheet-trigger"
    >
      Open Sheet
    </button>
    <div
      data-testid="sheet-portal"
    >
      <div
        class="fixed inset-0 z-50 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0"
        data-testid="sheet-overlay"
      />
      <div
        class="fixed z-50 gap-4 bg-background p-6 shadow-lg transition ease-in-out data-[state=closed]:duration-300 data-[state=open]:duration-500 data-[state=open]:animate-in data-[state=closed]:animate-out inset-y-0 left-0 h-full w-3/4 border-r data-[state=closed]:slide-out-to-left data-[state=open]:slide-in-from-left sm:max-w-sm"
        data-testid="sheet-content"
      >
        <button
          class="absolute right-4 top-4 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none  disabled:pointer-events-none data-[state=open]:bg-secondary"
          data-testid="sheet-close-button"
        >
          <span
            class="sr-only"
          >
            Close
          </span>
        </button>
        <div
          class="flex flex-col space-y-2 text-center sm:text-left"
        >
          <div
            class="text-lg font-semibold text-foreground"
            data-testid="sheet-title"
          >
            Sheet Title
          </div>
          <div
            class="text-sm text-muted-foreground"
            data-testid="sheet-description"
          >
            Sheet Description
          </div>
        </div>
        <div>
          Content
        </div>
        <div
          class="flex flex-col-reverse sm:flex-row sm:justify-end sm:space-x-2"
        >
          <button
            data-testid="sheet-close-button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Sheet Component > Sheet Snapshots > should match snapshot with default props 1`] = `
<div>
  <div
    data-testid="sheet-root"
  >
    <button
      data-testid="sheet-trigger"
    >
      Open Sheet
    </button>
    <div
      data-testid="sheet-portal"
    >
      <div
        class="fixed inset-0 z-50 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0"
        data-testid="sheet-overlay"
      />
      <div
        class="fixed z-50 gap-4 bg-background p-6 shadow-lg transition ease-in-out data-[state=closed]:duration-300 data-[state=open]:duration-500 data-[state=open]:animate-in data-[state=closed]:animate-out inset-y-0 right-0 h-full w-3/4 border-l data-[state=closed]:slide-out-to-right data-[state=open]:slide-in-from-right sm:max-w-sm"
        data-testid="sheet-content"
      >
        <button
          class="absolute right-4 top-4 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none  disabled:pointer-events-none data-[state=open]:bg-secondary"
          data-testid="sheet-close-button"
        >
          <span
            class="sr-only"
          >
            Close
          </span>
        </button>
        <div
          class="flex flex-col space-y-2 text-center sm:text-left"
        >
          <div
            class="text-lg font-semibold text-foreground"
            data-testid="sheet-title"
          >
            Sheet Title
          </div>
          <div
            class="text-sm text-muted-foreground"
            data-testid="sheet-description"
          >
            Sheet Description
          </div>
        </div>
        <div>
          Content
        </div>
        <div
          class="flex flex-col-reverse sm:flex-row sm:justify-end sm:space-x-2"
        >
          <button
            data-testid="sheet-close-button"
          >
            Close
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
